### ISSUE TITLE
Implementation of AutoDIAL: Automatic DomaIN Aignment Layers

### ISSUE THREAD

**Issue by Lan1991Xu** [domainadaptation/salad#21]
Hi. Thanks for sharing code. I just check:
class salad.layers.da.AutoAlign2d. It seems using original Batchnorm2D, any misunderstanding here.

**Comment by stes** [comment]
Hi @Lan1991Xu , thanks for flagging this, it is indeed not yet implemented, but it should be straightforward to do so. A similar (or related) functionality (conditional batch norm) was used in the conditional model used in my [experiments here](https://github.com/stes/nips2018-continual). I will think about how to best integrate this in the correct way into salad.

**Comment by Lan1991Xu** [comment]
Thanks.Do you means conditional batch norm here.   https://github.com/domainadaptation/salad/blob/190e3507d111967c61a8973fcf15d1a9c4e728ce/salad/models/digits/ensemble.py.  By the I also check the conditional model you mentioned in the link. It seems it has submit to NIPS workshop. Have you publish on the Arxiv? it would be good I can get access for the paper to better understand. Currently my understand is : it using compute_linear in test after we trained the model.  To implement AutoAlign2d, we need  calculate mean and variance of each BN by a linear transferform  spastic information between the source and target. Is this understanding right. Thanks.

**Comment by stes** [comment]
> Thanks.Do you means conditional batch norm here.

Yes, that would be an example on how to use it. Another one (in the context of ResNet for larger images) can be found [here](https://github.com/domainadaptation/salad/blob/190e3507d111967c61a8973fcf15d1a9c4e728ce/salad/models/resnet.py). 

They way it is implemented so far is by extending the `forward` function by a domain param, hence the function signature `forward(self, x, d)`. Depending on `d`, different batch norm statistics will be updated and a different affine transform is used. This functionality is however not yet well integrated into the use with all the implemented domain adaptation algorithms.

> By the I also check the conditional model you mentioned in the link. It seems it has submit to NIPS workshop. Have you publish on the Arxiv?

We don't currently have a pre-print available, but I will link the paper once we integrated feedback from the review process, that should be sometime mid-November.

> To implement AutoAlign2d, we need calculate mean and variance of each BN by a linear transferform spastic information between the source and target. Is this understanding right. Thanks.

AFAIK, the layer was first proposed [here](https://arxiv.org/pdf/1704.08082.pdf). To break it down:

Regarding the loss functions, the authors use entropy minimization on the target domain, which is already implemented in `salad`:

![image](https://user-images.githubusercontent.com/727984/47844097-c2e1dd00-dd7e-11e8-92af-3ab3040f24dc.png)

Regarding the layers, you are right (if I got your comment right):
![image](https://user-images.githubusercontent.com/727984/47844231-181dee80-dd7f-11e8-9540-c1362f3da3c7.png)

where the different `q` are mixing distributions:

![image](https://user-images.githubusercontent.com/727984/47844251-2b30be80-dd7f-11e8-9a3d-353633ab022d.png)

So actually, the algorithm can be straightforwardly implemented, I will see if I find some time in the near future.
